meta-mender-core/classes/mender-helpers.bbclass

```diff
@@ -242,12 +242,16 @@ def mender_get_data_part_num(d):
     n = 3
     boot_part_size = d.getVar('MENDER_BOOT_PART_SIZE_MB')
     swap_part_size = d.getVar('MENDER_SWAP_PART_SIZE_MB')
+    extra_parts = d.getVar('MENDER_EXTRA_PARTS')
     if (boot_part_size and boot_part_size != '0'): n += 1
     if (swap_part_size and swap_part_size != '0'): n += 1
+    if extra_parts: n += 1
 
     #is an msdos extended partion going to be required
     if n <= 4: return n
     if mender_is_msdos_ptable_image(d): n += 1
+    # extra partitions are added after data partition
+    if extra_parts: n -= 1
```

Comment:
Is this correct? Even if the data partition has extra partitions following it, it would keep the same index, right?

Reply:
Yes it is correct. In case we enable extra partition we need to add + 1 and in my case we have boot partition also it add +1 so we have 5. check if n <= 4 is false and then  mender_is_msdos_ptable_image add n+1 which is not right because then data will be mount on 6 and it should be on 5 (as 4 is extended partition). Point is that other partitions are added after data so thus -1.

Reply:
I'n not convinced this is the correct way to go. mender_get_data_part_num() is dedicated to getting the data partition index and should not need to know nothing about  extra partitions if they always go after the data partition.  if extra partitions added, the data partition index shouldn't need to change. Later code for creating partitions should take result of mender_get_data_part_num() and calculate their indexes accordingly. :)

Reply:
IMO it must be influenced because extra partitions maybe require to create extended partition. As you describe it I have it implemented and it was commented so I dropped it but certainly some action in this method (or other) must be done. If I drop it from mender_get_data_part_num without knowing about extra partitions then /data will be mounted on e.g. /dev/mmcblk0p4 which is not correct. So based on my testing this is correct solution :) (not saying it's elegant way). Let's think about proper one :+1:

Reply:
Is it not broken if both boot_part_size and swap_part_size are not declared resulting in data part 4 being returned when it should be 3?

Reply:
@dellgreen  yes this is right. I'll try to check it and fix for all configs.

Reply:
ok created this : https://pastebin.com/hAYV49gG to verify fix (oneliner only) and it looks like now it should be OK. Any other comments to this function? I also realize that would need some helper because if board don't have boot part and also not swap and user define 2 extra partition code in updating fstab is not correct in this case.

Reply:
@kacf looks good to me. One small nitpick which I see is:
```
    #is an msdos extended partion going to be required
-    if n <= 4: return n
-    if mender_is_msdos_ptable_image(d): n += 1
-    return n
+    if parts_num > 4 and mender_is_msdos_ptable_image(d):
+        parts_num += 1 <----- we don't need this line IMO
+        if data_pos >= 4:
+            data_pos += 1
+    return data_pos
```

Reply:
Well spotted, feel free to remove it in your update.

Reply:
Are you saying this line is not needed because its wrong, or because its redundant? 

if redundant then its probably worth keeping this line in for any code that may be added to this function in the future that may depend on an accurate parts_num calculation.

Reply:
Line is redundant because variable is not used anywhere after that point. Function should be used as name says for return data offset. For partition number I'll probably create new helper which return number of partitions in total (used for fstab update)